myia/opt/lib.py

```diff
@@ -189,6 +189,38 @@ def bubble_op_tuple_binary(optimizer, node, equiv):
 )
 
 
+@pattern_replacer(P.array_map, G, Xs)
+def unfuse_composite(optimizer, node, equiv):
+    """Tranform array_map on a graph to a graph of array_maps.
+
+    This must be applied to scalar-only graphs.
+    """
+    from ..grad import GraphRemapper
+
+    # This has to be defined inline because of circular imports
+    class UnfuseRemapper(GraphRemapper):
+        def __init__(self, g):
+            super().__init__('unfused', g.graphs_used.keys() | {g})
```

Comment:
You need to implement `gen_constant`, otherwise you'll generate code like `array_map(scalar_sub, xs, 1)`, but what you need to generate is something along the lines of `array_map(scalar_sub, xs, distribute(scalar_to_array(1), shape(xs)))`. This issue will happen with the gradient of `pow`.

Also, you can use the same trick on free variables (implement `gen_fv`). This would allow some closures to be passed to `array_map`.

Reply:
Ok.  I should also add some tests for that.

Reply:
implementing `gen_constant` turns out to be more complex than necessary since I have the primitives to worry about and I don't know the proper shape for the constant at that point.

Reply:
The shape can be found in `xs[0].shape` in the remapper's outer scope, but sure.